src/codegen/codegen_neuron_cpp_visitor.cpp

```diff
@@ -443,49 +538,78 @@ void CodegenNeuronCppVisitor::print_mechanism_register() {
+                                             : std::string{});
+    }
+    for (auto& var: codegen_float_variables) {
+        const auto& name = var->get_name();
+        // auto type = get_range_var_float_type(var);
+        // auto qualifier = is_constant_variable(name) ? "const " : "";
+        printer->fmt_line("double* {}{};", name, value_initialize);
+    }
+    for (auto& var: codegen_int_variables) {
+        const auto& name = var.symbol->get_name();
+        if (var.is_index || var.is_integer) {
+            auto qualifier = var.is_constant ? "const " : "";
+            printer->fmt_line("{}{}* {}{};", qualifier, int_type, name, value_initialize);
         } else {
-            printer->add_line("#define ",
-                              float_var->get_name(),
-                              "(id) _ml->template fpfield<",
-                              std::to_string(i),
-                              ">(id)");
+            auto qualifier = var.is_constant ? "const " : "";
+            auto type = var.is_vdata ? "void*" : default_float_data_type();
+            printer->fmt_line("{}{}* {}{};", qualifier, type, name, value_initialize);
         }
     }
+
+    // printer->fmt_line("{}* {}{};",
+    //                   global_struct(),
+    //                   naming::INST_GLOBAL_MEMBER,
+    //                   print_initializers ? fmt::format("{{&{}}}", global_struct_instance())
+    //                                      : std::string{});
+    printer->pop_block(";");
 }
 
 
-/// TODO: Edit for NEURON
 void CodegenNeuronCppVisitor::print_global_function_common_code(BlockType type,
                                                                 const std::string& function_name) {
-    return;
+    std::string method = function_name.empty() ? compute_method_name(type) : function_name;
+    std::string args =
+        "_nrn_model_sorted_token const& _sorted_token, NrnThread* _nt, Memb_list* _ml_arg, int "
+        "_type";
+    printer->fmt_push_block("void {}({})", method, args);
+
+    printer->add_line("_nrn_mechanism_cache_range _lmr{_sorted_token, *_nt, *_ml_arg, _type};");
+    printer->fmt_line("auto inst = make_instance_{}(_lmr);", info.mod_suffix);
+    printer->add_line("auto nodecount = _ml_arg->nodecount;");
 }
 
 
 void CodegenNeuronCppVisitor::print_nrn_init(bool skip_init_check) {
-    codegen = true;
-    printer->add_newline(2);
-    printer->add_line("/** initialize channel */");
+    print_global_function_common_code(BlockType::Initial);
 
-    printer->fmt_line(
-        "static void {}(_nrn_model_sorted_token const& _sorted_token, NrnThread* _nt, Memb_list* "
-        "_ml_arg, int _type) {{}}",
-        method_name(naming::NRN_INIT_METHOD));
+    printer->push_block("for (int id = 0; id < nodecount; id++)");
+    print_initial_block(info.initial_node);
+    printer->pop_block();
 
-    codegen = false;
+    printer->pop_block();
+}
+
+void CodegenNeuronCppVisitor::print_initial_block(const InitialBlock* node) {
```

Comment:
Should be moved above `print_global_function_common_code`. Let's try to keep the order of declaration of functions between `CodegenCoreneuronCppVisitor` and `CodegenNeuronCppVisitor` the same to easily diff them

Reply:
done